### PR TITLE
Fix `MathesarErrorMessageMixin` to convert exception objects

### LINKED ISSUE #1228: Internal server error after column type is converted from text to boolean with display options
## To Reproduce
* Create a new text column
* Patch the column from text to boolean with display options.
  ```javascript
  {
    type: 'boolean',
    type_options: {},
    display_options: {
      input: 'dropdown',
      custom_labels: {
        TRUE: 'TRUE'
        FALSE: 'FALSE'
      }
    }
  }
  ```
* Notice that the request succeeds.
* Send a subsequent request to get all columns of the table. Notice the 500 error.
* Try loading the frontend, a debug page with the error message appears.
* **This does not happen always, happens intermittently and I end up having to drop the table directly in the db inorder to get things working again.**

## Error message:
```
[{"code":4999,"message":"\"Got KeyError when attempting to get a value for field `input` on serializer `BooleanDisplayOptionSerializer`.\\nThe serializer field might be named incorrectly and not match any attribute or key on the `dict` instance.\\nOriginal exception text was: 'input'.\"","field":null,"detail":null}]
```

## Error log:
```
Environment:


Request Method: GET
Request URL: http://localhost:8000/mathesar_tables/1/?t=W1tdLG51bGxd

Django Version: 3.1.14
Python Version: 3.9.8
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'rest_framework',
 'django_filters',
 'django_property_filter',
 'mathesar']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware']



Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/rest_framework/fields.py", line 457, in get_attribute
    return get_attribute(instance, self.source_attrs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/fields.py", line 95, in get_attribute
    instance = instance[attr]

During handling of the above exception ('input'), another exception occurred:
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/code/mathesar/views.py", line 109, in schema_home
    'common_data': get_common_data(request, database, schema)
  File "/code/mathesar/views.py", line 55, in get_common_data
    'tables': get_table_list(request, schema),
  File "/code/mathesar/views.py", line 35, in get_table_list
    return table_serializer.data
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 745, in data
    ret = super().data
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 246, in data
    self._data = self.to_representation(self.instance)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 663, in to_representation
    return [
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 664, in <listcomp>
    self.child.to_representation(item) for item in iterable
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 515, in to_representation
    ret[field.field_name] = field.to_representation(attribute)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 663, in to_representation
    return [
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 664, in <listcomp>
    self.child.to_representation(item) for item in iterable
  File "/code/mathesar/api/serializers/columns.py", line 67, in to_representation
    return super().to_representation(instance)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 515, in to_representation
    ret[field.field_name] = field.to_representation(attribute)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 502, in to_representation
    attribute = field.get_attribute(instance)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/fields.py", line 490, in get_attribute
    raise type(exc)(msg)

Exception Type: KeyError at /mathesar_tables/1/
Exception Value: "Got KeyError when attempting to get a value for field `input` on serializer `BooleanDisplayOptionSerializer`.\nThe serializer field might be named incorrectly and not match any attribute or key on the `dict` instance.\nOriginal exception text was: 'input'."
```

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #1228 

Adds support for raising Custom Exception with an object as error body inside DRF serializer. 

**Technical details**
DRF serializers support exceptions in any of the following formats, error string, list of error strings or {'field': [error strings]}. But our custom validation exceptions use objects as the error body, which gets treated as `{'field': [error strings]}` by DRF serializer. This PR adds support to bypass the conversion if the exception body already has a proper format according to our spec.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: mathesar/api/exceptions/mixins.py [hunk 1]
```diff
@@ -19,6 +19,13 @@ def build_pretty_errors(self, errors, serializer=None):
         2. Add field to the pretty exception body if raised by field validation method
         """
         pretty = []
+        if self.is_pretty(errors):
+            # DRF serializers supports error any of the following format, error string, list of error strings or {'field': [error strings]}
+            # Since our exception is an object instead of a string, the object properties are mistaken to be fields of a serializer,
+            # and it gets converted into {'field': [error strings]} by DRF
+            # We need to convert it to dictionary of list of object and return it instead of passing it down the line
+            scalar_errors = dict(map(lambda item: (item[0], item[1][0] if type(item[1]) == list else item[1]), errors.items()))
```
Comment (kgodey) [inline mathesar/api/exceptions/mixins.py:27]: @silentninja This logic is a little hard to follow, could you split it into multiple lines?

**Review by kgodey — CHANGES_REQUESTED** [review]

**Comment by silentninja** [comment]
Closing this in favour of #1251 as that is the actual root cause of the issue #1228
